target/linux/generic/backport-5.4/700-v5.5-net-core-allow-fast-GRO-for-skbs-with-Ethernet-heade.patch

```diff
@@ -66,13 +66,43 @@ Signed-off-by: David S. Miller <davem@davemloft.net>
 
 --- a/net/core/dev.c
 +++ b/net/core/dev.c
-@@ -5404,8 +5404,7 @@ static void skb_gro_reset_offset(struct
+@@ -5395,7 +5395,7 @@ static struct list_head *gro_list_prepar
+ 	return head;
+ }
+ 
+-static void skb_gro_reset_offset(struct sk_buff *skb)
++static inline void skb_gro_reset_offset(struct sk_buff *skb, u32 nhoff)
+ {
+ 	const struct skb_shared_info *pinfo = skb_shinfo(skb);
+ 	const skb_frag_t *frag0 = &pinfo->frags[0];
+@@ -5404,10 +5404,9 @@ static void skb_gro_reset_offset(struct
  	NAPI_GRO_CB(skb)->frag0 = NULL;
  	NAPI_GRO_CB(skb)->frag0_len = 0;
  
 -	if (skb_mac_header(skb) == skb_tail_pointer(skb) &&
 -	    pinfo->nr_frags &&
 +	if (!skb_headlen(skb) && pinfo->nr_frags &&
- 	    !PageHighMem(skb_frag_page(frag0))) {
+ 	    !PageHighMem(skb_frag_page(frag0)) &&
+-	    (!NET_IP_ALIGN || !(skb_frag_off(frag0) & 3))) {
++	    (!NET_IP_ALIGN || !((skb_frag_off(frag0) + nhoff) & 3))) {
  		NAPI_GRO_CB(skb)->frag0 = skb_frag_address(frag0);
  		NAPI_GRO_CB(skb)->frag0_len = min_t(unsigned int,
+ 						    skb_frag_size(frag0),
+@@ -5640,7 +5639,7 @@ gro_result_t napi_gro_receive(struct nap
+ 	skb_mark_napi_id(skb, napi);
+ 	trace_napi_gro_receive_entry(skb);
+ 
+-	skb_gro_reset_offset(skb);
++	skb_gro_reset_offset(skb, 0);
+ 
+ 	ret = napi_skb_finish(napi, skb, dev_gro_receive(napi, skb));
+ 	trace_napi_gro_receive_exit(ret);
+@@ -5733,7 +5732,7 @@ static struct sk_buff *napi_frags_skb(st
+ 	napi->skb = NULL;
+ 
+ 	skb_reset_mac_header(skb);
+-	skb_gro_reset_offset(skb);
++	skb_gro_reset_offset(skb, hlen);
```

Comment:
Why is this hlen needed here?
Kernel 5.10.32 contains also both patches, but just ```skb_gro_reset_offset(skb)``` is called here.

Reply:
@solbjorn - Can you address this question?

Reply:
@hauke, @graysky2, we decided to also include this upstream fix: https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=7ad18ff6449cbd6beb26b53128ddf56d2685aa93
It will land mainline soon, the reason why it's important is because Eric Dumazet's check for NET_IP_ALIGN causes huge slowdowns on drivers which use `napi_gro_frags()`.

Reply:
Thank you for the link.
I would prefer if you add this as an extra patch into the backports-5.4 folder and not squash multiple upstream commits together here.

Reply:
@hauke - I am happy to do so (@solbjorn actually provided them to me as two discrete patches).  I am but am ignorant as to how to add a patch.  Do I simply place the new file in `target/linux/generic/backport-5.4`?

Reply:
@hauke - Done, please see the new commit force pushed.